### PR TITLE
Implement Number.parseFloat, Number.parseInt

### PR DESCRIPTION
**Pull request by orkvi**
…er.parseInt

I copied Global object parseFloat, parseInt fuctions to the Number and Intrinsic objects,
because in these two objects the functions work the same way.

JerryScript-DCO-1.0-Signed-off-by: Virag Orkenyi orkvi@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 1]
```diff
@@ -64,6 +80,109 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+      break;
+    }
+    case ECMA_INTRINSIC_PARSE_INT:
+    {
+      ret_value = ecma_builtin_intrinsic_object_parse_int (input_start_p, input_size, routine_arg_2);
+      break;
+    }
+    case ECMA_INTRINSIC_ARRAY_PROTOTYPE_VALUES:
+    {
+      ret_value = ecma_builtin_intrinsic_array_prototype_values (this_arg);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: Tostring , and get string buffer operation should not be applied for this method.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c [hunk 2]
```diff
@@ -730,7 +730,7 @@ ecma_builtin_json_parse_buffer (const lit_utf8_byte_t * str_start_p, /**< String
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
-static ecma_value_t
+ecma_value_t
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: What is the reason of this change?

File: tests/jerry/es2015/number-methods.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+assert (Number.parseInt ('18528769') === 18528769);
+assert (Number.parseFloat ('1.2e3') === 1.2e3);
+
+Number.parseInt.prop = 'apple';
+assert (parseInt.prop === 'apple');
+
+
```
Comment (rerobika) [inline tests/jerry/es2015/number-methods.js]: You should check that the global and the number built-in methods are the same.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c (hunk 2 above)
Comment (orkvi) [reply jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: Nothing. It was a mistake. I didn't want to add this change.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.c [hunk 4]
```diff
@@ -302,180 +142,10 @@ ecma_builtin_global_object_parse_int (const lit_utf8_byte_t *string_buff, /**< r
 static ecma_value_t
 ecma_builtin_global_object_parse_float (const lit_utf8_byte_t *string_buff, /**< routine's first argument's
                                                                              *   string buffer */
-                                        lit_utf8_size_t string_buff_size) /**< routine's first argument's
+                                           lit_utf8_size_t string_buff_size) /**< routine's first argument's
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.c]: `ecma_builtin_global_object_parse_float`  is only a jumber function now, so it can be removed and `ecma_number_parse_float ` can be called directly. (same for parseInt)

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 5]
```diff
@@ -64,6 +80,95 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+ * The Intrinsic object's 'parseInt' routine
+ *
+ * See also:
+ *          ECMA-262 v5, 15.1.2.2
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_intrinsic_object_parse_int (const lit_utf8_byte_t *string_buff,
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: Ditto, the jumper function can be removed.

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 6]
```diff
@@ -64,6 +80,95 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+                                         ecma_value_t this_arg, /**< 'this' argument value */
+                                         const ecma_value_t arguments_list_p[], /**< list of arguments
+                                                                              *   passed to routine */
+                                         ecma_length_t arguments_number) /**< length of arguments' list */
+{
+  JERRY_UNUSED (arguments_number);
+
+  ecma_value_t routine_arg_1 = arguments_list_p[0];
+  ecma_value_t routine_arg_2 = arguments_list_p[1];
+  ecma_value_t ret_value = 0;
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: That's an invalid assignment.

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 7]
```diff
@@ -64,6 +80,95 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+
+  if (builtin_routine_id == ECMA_INTRINSIC_ARRAY_PROTOTYPE_VALUES)
+  {
+    ret_value = ecma_builtin_intrinsic_array_prototype_values (this_arg);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: you can return directly with the result of this function call.

File: jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h [hunk 8]
```diff
@@ -89,5 +89,8 @@ ROUTINE (LIT_MAGIC_STRING_IS_FINITE, ecma_builtin_number_object_is_finite, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_NAN, ecma_builtin_number_object_is_nan, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_INTEGER, ecma_builtin_number_object_is_integer, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_SAFE_INTEGER, ecma_builtin_number_object_is_safe_integer, 1, 1)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_FLOAT, LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_PROPERTY_FIXED)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_INT, LIT_MAGIC_STRING_PARSE_INT, ECMA_PROPERTY_FIXED)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h]: These methods are configurable and writable.

File: tests/jerry/es2015/number-methods.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+// limitations under the License.
+
+assert (Number.parseInt ('18528769') === 18528769);
+assert (Number.parseFloat ('1.2e3') === 1.2e3);
+assert (parseInt ('18528769') === 18528769);
+assert (parseFloat ('1.2e3') === 1.2e3);
+
+assert(Number.parseInt===parseInt);
+
+assert(Number.parseFloat===parseFloat);
```
Comment (rerobika) [inline tests/jerry/es2015/number-methods.js]: spaces around `===`

File: tests/jerry/es2015/number-methods.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+
```
Comment (rerobika) [inline tests/jerry/es2015/number-methods.js]: Too many extra new lines.

File: jerry-core/ecma/base/ecma-helpers-number.c [hunk 11]
```diff
@@ -12,9 +12,26 @@
...
+#include "ecma-try-catch-macro.h"
+#include "jrt.h"
+#include "lit-char-helpers.h"
+#include "lit-magic-strings.h"
+#include "lit-strings.h"
+#include "vm.h"
+#include "jcontext.h"
+#include "jrt-libc-includes.h"
+#include "jrt-bit-fields.h"
+/*#include "ecma-globals.h"
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-number.c]: Are you sure that all these new includes are needed?

File: jerry-core/ecma/base/ecma-helpers.h [hunk 12]
```diff
@@ -347,6 +347,11 @@ ecma_number_t ecma_number_get_prev (ecma_number_t num);
 ecma_number_t ecma_number_get_next (ecma_number_t num);
 ecma_number_t ecma_number_trunc (ecma_number_t num);
 ecma_number_t ecma_number_calc_remainder (ecma_number_t left_num, ecma_number_t right_num);
+ecma_value_t ecma_number_parse_int (const lit_utf8_byte_t *string_buff,
+                                     lit_utf8_size_t string_buff_size,
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers.h]: misaligned arguments.

File: jerry-core/ecma/base/ecma-helpers-number.c [hunk 13]
```diff
@@ -12,9 +12,20 @@
...
+#include "ecma-exceptions.h"
+#include "ecma-gc.h"
 #include "ecma-globals.h"
 #include "ecma-helpers.h"
+#include "ecma-try-catch-macro.h"
+//#include "jrt.h"
+#include "lit-char-helpers.h"
+#include "lit-magic-strings.h"
+#include "lit-strings.h"
+//#include "vm.h"
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-number.c]: Please reduce the number of new includes, most of them are unnecessary.

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 14]
```diff
@@ -64,6 +80,60 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+    {
+      ret_value = ecma_number_parse_float (string_buff,
+                                           string_buff_size);
+    }
+    ECMA_FINALIZE_UTF8_STRING (string_buff, string_buff_size);
+    ecma_deref_ecma_string (str_p);
+  }
+
+  return ret_value;
+
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: no new line is needed.

File: jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h [hunk 15]
```diff
@@ -89,5 +89,8 @@ ROUTINE (LIT_MAGIC_STRING_IS_FINITE, ecma_builtin_number_object_is_finite, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_NAN, ecma_builtin_number_object_is_nan, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_INTEGER, ecma_builtin_number_object_is_integer, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_SAFE_INTEGER, ecma_builtin_number_object_is_safe_integer, 1, 1)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_FLOAT, LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_INT, LIT_MAGIC_STRING_PARSE_INT, ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
 #endif /* ENABLED (JERRY_ES2015) */
+
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h]: Ditto.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h [hunk 16]
```diff
@@ -248,17 +248,20 @@ OBJECT_VALUE (LIT_MAGIC_STRING_DATAVIEW_UL,
...
 ROUTINE (LIT_MAGIC_STRING_ENCODE_URI, ECMA_GLOBAL_ENCODE_URI, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_ENCODE_URI_COMPONENT, ECMA_GLOBAL_ENCODE_URI_COMPONENT, 1, 1)
+#if ENABLED (JERRY_ES2015)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_FLOAT, LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_INT, LIT_MAGIC_STRING_PARSE_INT, ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
+#else /* !ENABLED (JERRY_ES2015)*/
+ROUTINE (LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_GLOBAL_PARSE_FLOAT, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_PARSE_INT, ECMA_GLOBAL_PARSE_INT, 2, 2)
-
+#endif /* ENABLED (JERRY_ES2015)*/
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h]: ditto. Also please do not remove the line above.

File: jerry-core/ecma/base/ecma-helpers-number.c [hunk 17]
```diff
@@ -12,9 +12,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-
```
Comment (rerobika) [inline jerry-core/ecma/base/ecma-helpers-number.c]: Please leave this line after the license block.

File: jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h [hunk 18]
```diff
@@ -248,17 +248,20 @@ OBJECT_VALUE (LIT_MAGIC_STRING_DATAVIEW_UL,
 
 /* Routine properties:
  *  (property name, C routine name, arguments number or NON_FIXED, value of the routine's length property) */
-
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-global.inc.h]: Ditto, leave this line.

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 19]
```diff
@@ -16,12 +16,28 @@
...
+ * This object has a custom dispatch function.
+ */
+#define BUILTIN_CUSTOM_DISPATCH
+
+/**
+ * List of built-in routine identifiers.
+ */
+enum
+{
+  ECMA_INTRINSIC_PARSE_FLOAT = ECMA_BUILTIN_ID__COUNT,
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: I'd add the first entry to be consistent with the other custom dispatchers like::

```
ECMA_INTRINSIC_ROUTINE_START = ECMA_BUILTIN_ID__COUNT - 1,
```
after that `ECMA_INTRINSIC_PARSE_FLOAT`, is incremented automatically. ,

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 20]
```diff
@@ -64,6 +80,59 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+
+    if (builtin_routine_id == ECMA_INTRINSIC_PARSE_INT)
+    {
+      ret_value = ecma_number_parse_int (string_buff,
+                                         string_buff_size,
+                                         routine_arg_2);
+    }
+    else
+    {
+      ret_value = ecma_number_parse_float (string_buff,
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: Please assert that the `builtin_id` must be `ECMA_INTRINSIC_PARSE_FLOAT`

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c [hunk 21]
```diff
@@ -64,6 +80,59 @@ ecma_builtin_intrinsic_array_prototype_values (ecma_value_t this_value) /**< thi
...
+                                                                              *   passed to routine */
+                                         ecma_length_t arguments_number) /**< length of arguments' list */
+{
+  JERRY_UNUSED (arguments_number);
+
+  ecma_value_t routine_arg_1 = arguments_list_p[0];
+  ecma_value_t routine_arg_2 = arguments_list_p[1];
+  ecma_value_t ret_value = jerry_create_null ();
+
+  if (builtin_routine_id == ECMA_INTRINSIC_ARRAY_PROTOTYPE_VALUES)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.c]: This method is only present is ES2015 profile.

File: jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.inc.h [hunk 22]
```diff
@@ -64,8 +64,10 @@ SYMBOL_VALUE (LIT_GLOBAL_SYMBOL_TO_STRING_TAG,
...
 SYMBOL_VALUE (LIT_GLOBAL_SYMBOL_UNSCOPABLES,
               LIT_MAGIC_STRING_UNSCOPABLES)
-
-ROUTINE (LIT_INTERNAL_MAGIC_STRING_ARRAY_PROTOTYPE_VALUES, ecma_builtin_intrinsic_array_prototype_values, 0, 0)
-
+/* Routine properties:
+ *  (property name, C routine name, arguments number or NON_FIXED, value of the routine's length property) */
+ROUTINE (LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_INTRINSIC_PARSE_FLOAT, 1, 1)
+ROUTINE (LIT_MAGIC_STRING_PARSE_INT, ECMA_INTRINSIC_PARSE_INT, 2, 2)
+ROUTINE (LIT_INTERNAL_MAGIC_STRING_ARRAY_PROTOTYPE_VALUES, ECMA_INTRINSIC_ARRAY_PROTOTYPE_VALUES, 0, 0)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-intrinsic.inc.h]: Ditto, this routine should be guarded.

File: jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h [hunk 23]
```diff
@@ -89,5 +89,7 @@ ROUTINE (LIT_MAGIC_STRING_IS_FINITE, ecma_builtin_number_object_is_finite, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_NAN, ecma_builtin_number_object_is_nan, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_INTEGER, ecma_builtin_number_object_is_integer, 1, 1)
 ROUTINE (LIT_MAGIC_STRING_IS_SAFE_INTEGER, ecma_builtin_number_object_is_safe_integer, 1, 1)
+INTRINSIC_PROPERTY (LIT_MAGIC_STRING_PARSE_FLOAT, LIT_MAGIC_STRING_PARSE_FLOAT, ECMA_PROPERTY_CONFIGURABLE_WRITABLE)
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-number.inc.h]: Also, guard these methods.
